File: javascript/json_batch_processor.js
#!/usr/bin/env node
// json_batch_processor.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_batch_processor.py
// auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_batch_processor.py
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

/**
 * Batch JSON Processor - Verarbeitet mehrere JSON-Dateien oder JSON-Lines (NDJSON).
 */

const fs = require('fs');
const path = require('path');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

// Simple JSON repair function (basic implementation)
function repairJson(jsonString) {
  // Entferne führende/trailing whitespace
  jsonString = jsonString.trim();
  
  // Ersetze einfache Anführungszeichen durch doppelte (wenn nicht escaped)
  jsonString = jsonString.replace(/'/g, '"');
  
  // Entferne trailing Kommas
  jsonString = jsonString.replace(/,\s*([\]}])/g, '$1');
  
  return jsonString;
}

// Parse JSON with optional repair
function parseJson(content, repair = true) {
  try {
    return JSON.parse(content);
  } catch (e) {
    if (repair) {
      try {
        const repaired = repairJson(content);
        return JSON.parse(repaired);
      } catch (repairError) {
        throw new Error(`JSON parse error: ${e.message}`);
      }
    } else {
      throw new Error(`JSON parse error: ${e.message}`);
    }
  }
}

// BatchResult Klasse
class BatchResult {
  constructor(index, source, success, data = null, error = null) {
    this.index = index;
    this.source = source;
    this.success = success;
    this.data = data;
    this.error = error;
  }

  toDict() {
    return {
      index: this.index,
      source: this.source,
      success: this.success,
      data: this.data,
      error: this.error
    };
  }
}

// Liest JSON-Lines (NDJSON) Datei Zeile für Zeile
function* readJsonl(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');
  
  for (let lineNum = 0; lineNum < lines.length; lineNum++) {
    const line = lines[lineNum].trim();
    if (!line) continue;
    
    try {
      yield JSON.parse(line);
    } catch (e) {
      yield new BatchResult(
        lineNum + 1,
        `${filePath}:${lineNum + 1}`,
        false,
        null,
        `JSON decode error: ${e.message}`
      );
    }
  }
}

// Verarbeitet eine Liste von Inputs parallel
async function processBatch(inputs, processor, maxWorkers = 4) {
  const results = [];
  const workers = [];
  const workerPromises = [];
  
  // Node.js unterstützt keine echte Thread-Pools wie Python,
  // daher simulieren wir parallele Verarbeitung mit Promises
  const promises = inputs.map((inp, idx) => 
    Promise.resolve().then(() => processor(inp, idx))
  );
  
  // Warte auf alle Promises
  for (const promise of promises) {
    try {
      const result = await promise;
      results.push(result);
    } catch (e) {
      // Im Fehlerfall müssten wir den Index kennen, was hier schwierig ist
      results.push(new BatchResult(
        results.length,
        'unknown',
        false,
        null,
        `Unexpected error: ${e.message}`
      ));
    }
  }
  
  // Sortiere nach Index
  results.sort((a, b) => a.index - b.index);
  return results;
}

// Verarbeitet mehrere JSON-Dateien im Batch
async function processFileBatch(filePaths, repair = true, validateModel = null, maxWorkers = 4) {
  const processor = async (filePath, idx) => {
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      
      // In JS gibt es kein Pydantic, daher überspringen wir die Validierung
      const data = parseJson(content, repair);
      
      return new BatchResult(
        idx,
        filePath,
        true,
        data
      );
    } catch (e) {
      return new BatchResult(
        idx,
        filePath,
        false,
        null,
        e.message
      );
    }
  };
  
  return await processBatch(filePaths, processor, maxWorkers);
}

// Verarbeitet eine JSON-Lines Datei
function processJsonlFile(filePath, repair = true, validateModel = null) {
  const results = [];
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');
  
  for (let lineNum = 0; lineNum < lines.length; lineNum++) {
    const line = lines[lineNum].trim();
    if (!line) continue;
    
    try {
      // In JS gibt es kein Pydantic, daher überspringen wir die Validierung
      const data = parseJson(line, repair);
      
      results.push(new BatchResult(
        lineNum + 1,
        `${filePath}:${lineNum + 1}`,
        true,
        data
      ));
    } catch (e) {
      results.push(new BatchResult(
        lineNum + 1,
        `${filePath}:${lineNum + 1}`,
        false,
        null,
        e.message
      ));
    }
  }
  
  return results;
}

// Schreibt BatchResult-Liste als JSON-Lines
function writeJsonl(results, outputPath, onlySuccessful = true) {
  const fd = fs.openSync(outputPath, 'w');
  
  for (const result of results) {
    if (onlySuccessful && !result.success) continue;
    const jsonLine = JSON.stringify(result.toDict());
    fs.writeSync(fd, jsonLine + '\n');
  }
  
  fs.closeSync(fd);
}

// Hauptfunktion
async function main() {
  const args = process.argv.slice(2);
  const options = {
    inputs: [],
    jsonl: false,
    repair: true,
    workers: 4,
    output: null,
    summary: false
  };
  
  // Einfacher Argument Parser
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    
    if (arg === '--jsonl' || arg === '-l') {
      options.jsonl = true;
    } else if (arg === '--repair' || arg === '-r') {
      options.repair = true;
    } else if (arg === '--workers' || arg === '-w') {
      options.workers = parseInt(args[++i]);
    } else if (arg === '--output' || arg === '-o') {
      options.output = args[++i];
    } else if (arg === '--summary' || arg === '-s') {
      options.summary = true;
    } else if (!arg.startsWith('-')) {
      options.inputs.push(arg);
    }
  }
  
  if (options.inputs.length === 0) {
    console.error('ERROR: No input files provided');
    process.exit(1);
  }
  
  let allResults = [];
  
  if (options.jsonl) {
    // JSON-Lines Modus
    for (const inputPath of options.inputs) {
      const results = processJsonlFile(inputPath, options.repair);
      allResults = allResults.concat(results);
    }
  } else {
    // Standard JSON Batch
    allResults = await processFileBatch(
      options.inputs,
      options.repair,
      null,
      options.workers
    );
  }
  
  // Ausgabe
  const successful = allResults.filter(r => r.success).length;
  const failed = allResults.length - successful;
  
  if (options.summary) {
    console.log(`Processed: ${allResults.length}`);
    console.log(`Successful: ${successful}`);
    console.log(`Failed: ${failed}`);
  } else {
    for (const result of allResults) {
      if (result.success) {
        console.log(JSON.stringify(result.data));
      } else {
        console.error(`ERROR [${result.source}]: ${result.error}`);
      }
    }
  }
  
  // Optional: JSONL Output
  if (options.output) {
    writeJsonl(allResults, options.output, false);
    console.error(`\nResults written to: ${options.output}`);
  }
  
  // Exit code
  process.exit(failed === 0 ? 0 : 1);
}

if (require.main === module) {
  main().catch(e => {
    console.error(e);
    process.exit(1);
  });
}

module.exports = {
  BatchResult,
  readJsonl,
  processBatch,
  processFileBatch,
  processJsonlFile,
  writeJsonl,
  parseJson
};

File: perl5/json_batch_processor.pl
#!/usr/bin/perl
# json_batch_processor.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_batch_processor.py
# auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_batch_processor.py
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use JSON::PP;
use Getopt::Long;
use File::Spec;
use File::Basename;
use threads;
use Thread::Queue;

# JSON Batch Processor - Verarbeitet mehrere JSON-Dateien oder JSON-Lines (NDJSON)

# Globale Variablen
my $json = JSON::PP->new->utf8;
my $HAS_PYDANTIC = 0; # In Perl nicht direkt verfügbar, daher immer 0

# BatchResult-Klasse (simuliert)
package BatchResult {
    sub new {
        my ($class, $index, $source, $success, $data, $error) = @_;
        my $self = {
            index => $index,
            source => $source,
            success => $success,
            data => $data,
            error => $error
        };
        bless $self, $class;
        return $self;
    }
    
    sub to_dict {
        my ($self) = @_;
        return {
            index => $self->{index},
            source => $self->{source},
            success => $self->{success},
            data => $self->{data},
            error => $self->{error}
        };
    }
}

# Hauptpaket
package main;

# Funktion zum Lesen von JSON-Lines (NDJSON)
sub read_jsonl {
    my ($file_path) = @_;
    my @results;
    
    open my $fh, '<:encoding(UTF-8)', $file_path or die "Cannot open $file_path: $!";
    
    my $line_num = 0;
    while (my $line = <$fh>) {
        $line_num++;
        chomp $line;
        $line =~ s/^\s+|\s+$//g; # strip
        next if !$line;
        
        eval {
            my $data = $json->decode($line);
            push @results, $data;
        };
        if ($@) {
            push @results, BatchResult->new(
                $line_num,
                "$file_path:$line_num",
                0,
                undef,
                "JSON decode error: $@"
            );
        }
    }
    close $fh;
    return @results;
}

# Funktion zur parallelen Batch-Verarbeitung
sub process_batch {
    my ($inputs_ref, $processor_ref, $max_workers) = @_;
    $max_workers //= 4;
    
    my @results;
    my @threads;
    my $queue = Thread::Queue->new();
    my $result_queue = Thread::Queue->new();
    
    # Worker threads starten
    for my $i (1..$max_workers) {
        my $thr = threads->create(sub {
            while (defined(my $item = $queue->dequeue())) {
                my ($inp, $idx) = @$item;
                eval {
                    my $result = $processor_ref->($inp, $idx);
                    $result_queue->enqueue($result);
                };
                if ($@) {
                    my $result = BatchResult->new(
                        $idx,
                        "$inp",
                        0,
                        undef,
                        "Unexpected error: $@"
                    );
                    $result_queue->enqueue($result);
                }
            }
        });
        push @threads, $thr;
    }
    
    # Aufgaben in die Queue stellen
    for my $idx (0..$#{$inputs_ref}) {
        $queue->enqueue([$inputs_ref->[$idx], $idx]);
    }
    
    # Queue beenden
    $queue->end();
    
    # Ergebnisse sammeln
    for my $thr (@threads) {
        $thr->join();
    }
    
    # Ergebnisse aus der Queue holen
    while (defined(my $result = $result_queue->dequeue_nb())) {
        push @results, $result;
    }
    
    # Nach Index sortieren
    @results = sort { $a->{index} <=> $b->{index} } @results;
    return @results;
}

# Funktion zur Verarbeitung von JSON-Dateien im Batch
sub process_file_batch {
    my ($file_paths_ref, $repair, $validate_model, $max_workers) = @_;
    $repair //= 1;
    $max_workers //= 4;
    
    my $processor = sub {
        my ($path, $idx) = @_;
        eval {
            open my $fh, '<:encoding(UTF-8)', $path or die "Cannot open $path: $!";
            my $content = do { local $/; <$fh> };
            close $fh;
            
            my $data;
            if ($validate_model && $HAS_PYDANTIC) {
                # In Perl keine direkte Pydantic-Unterstützung
                $data = parse_json($content, $repair);
            } else {
                $data = parse_json($content, $repair);
            }
            
            return BatchResult->new(
                $idx,
                "$path",
                1,
                $data,
                undef
            );
        };
        if ($@) {
            my $error = $@;
            chomp $error;
            return BatchResult->new(
                $idx,
                "$path",
                0,
                undef,
                $error
            );
        }
    };
    
    return process_batch($file_paths_ref, $processor, $max_workers);
}

# Funktion zur Verarbeitung von JSON-Lines-Dateien
sub process_jsonl_file {
    my ($file_path, $repair, $validate_model) = @_;
    $repair //= 1;
    
    my @results;
    open my $fh, '<:encoding(UTF-8)', $file_path or die "Cannot open $file_path: $!";
    
    my $line_num = 0;
    while (my $line = <$fh>) {
        $line_num++;
        chomp $line;
        $line =~ s/^\s+|\s+$//g; # strip
        next if !$line;
        
        eval {
            my $data;
            if ($validate_model && $HAS_PYDANTIC) {
                # In Perl keine direkte Pydantic-Unterstützung
                $data = parse_json($line, $repair);
            } else {
                $data = parse_json($line, $repair);
            }
            
            push @results, BatchResult->new(
                $line_num,
                "$file_path:$line_num",
                1,
                $data,
                undef
            );
        };
        if ($@) {
            my $error = $@;
            chomp $error;
            push @results, BatchResult->new(
                $line_num,
                "$file_path:$line_num",
                0,
                undef,
                $error
            );
        }
    }
    close $fh;
    return @results;
}

# Funktion zum Schreiben von JSON-Lines
sub write_jsonl {
    my ($results_ref, $output_path, $only_successful) = @_;
    $only_successful //= 1;
    
    open my $fh, '>:encoding(UTF-8)', $output_path or die "Cannot open $output_path: $!";
    
    for my $result (@$results_ref) {
        if ($only_successful && !$result->{success}) {
            next;
        }
        my $dict = $result->to_dict();
        print $fh $json->encode($dict) . "\n";
    }
    close $fh;
}

# Funktion zum Parsen von JSON mit Reparatur (vereinfacht)
sub parse_json {
    my ($content, $repair) = @_;
    $repair //= 1;
    
    eval {
        return $json->decode($content);
    };
    if ($@ && $repair) {
        # Versuche einfache Reparaturen
        my $cleaned = $content;
        # Entferne trailing commas
        $cleaned =~ s/,(\s*[}\]])/$1/g;
        # Entferne leading commas
        $cleaned =~ s/(\s*)\,(?=\s*["{\[])/$1/g;
        
        eval {
            return $json->decode($cleaned);
        };
        if ($@) {
            die "JSON parsing failed: $@";
        }
    } elsif ($@) {
        die "JSON parsing failed: $@";
    }
}

# Hauptfunktion
sub main {
    my @inputs;
    my $jsonl = 0;
    my $repair = 1;
    my $workers = 4;
    my $output = '';
    my $summary = 0;
    
    GetOptions(
        "jsonl|l" => \$jsonl,
        "repair|r" => \$repair,
        "workers|w=i" => \$workers,
        "output|o=s" => \$output,
        "summary|s" => \$summary,
        "<>" => sub { push @inputs, @_; }
    ) or die "Falsche Optionen\n";
    
    if (!@inputs) {
        die "Keine Eingabedateien angegeben\n";
    }
    
    my @all_results;
    
    if ($jsonl) {
        # JSON-Lines Modus
        for my $input_path (@inputs) {
            my @results = process_jsonl_file($input_path, $repair);
            push @all_results, @results;
        }
    } else {
        # Standard JSON Batch
        my @file_paths = @inputs;
        @all_results = process_file_batch(\@file_paths, $repair, undef, $workers);
    }
    
    # Ausgabe
    my $successful = 0;
    for my $result (@all_results) {
        $successful++ if $result->{success};
    }
    my $failed = scalar(@all_results) - $successful;
    
    if ($summary) {
        print "Processed: " . scalar(@all_results) . "\n";
        print "Successful: $successful\n";
        print "Failed: $failed\n";
    } else {
        for my $result (@all_results) {
            if ($result->{success}) {
                print $json->encode($result->{data}) . "\n";
            } else {
                print STDERR "ERROR [$result->{source}]: $result->{error}\n";
            }
        }
    }
    
    # Optional: JSONL Output
    if ($output) {
        write_jsonl(\@all_results, $output, 0);
        print STDERR "\nResults written to: $output\n";
    }
    
    # Exit code
    exit($failed == 0 ? 0 : 1);
}

main() if !caller;
